File: src/hetzner_code/commands/pull.py
from __future__ import annotations

from pathlib import Path

import click

from hetzner_code import ssh_util, state
from hetzner_code.config import REMOTE_CODE_DIR
from hetzner_code.run import HcodeError


def pull(
    *,
    instance_name: str,
    remote_path: str,
    local_path: str | None,
    repo_name: str | None,
) -> None:
    instance = state.load(instance_name)

    resolved_remote = remote_path
    if repo_name:
        if not any(r.name == repo_name for r in instance.repos):
            available = ", ".join(r.name for r in instance.repos) or "(none)"
            raise HcodeError(
                f"'{repo_name}' isn't on '{instance_name}'. On it: {available}"
            )
        if not remote_path.startswith("/"):
            resolved_remote = f"{REMOTE_CODE_DIR}/{repo_name}/{remote_path}"

    resolved_local = (
        Path(local_path) if local_path else Path.cwd() / Path(remote_path).name
    )

    ssh_util.copy_from(
        instance.ip, resolved_remote, resolved_local, Path(instance.login_key_path)
    )
    click.echo(f"pulled {resolved_remote} -> {resolved_local}")

File: internal/commands/pull.go
package commands

import (
	"fmt"
	"os"
	"path"
	"path/filepath"
	"strings"

	"github.com/dikee/hetzner-code/internal/config"
	"github.com/dikee/hetzner-code/internal/run"
	"github.com/dikee/hetzner-code/internal/sshutil"
	"github.com/dikee/hetzner-code/internal/state"
)

// PullOptions are every flag `hcode pull` accepts.
type PullOptions struct {
	InstanceName string
	RemotePath   string
	LocalPath    string
	RepoName     string
}

// Pull runs `hcode pull`.
func Pull(opts PullOptions) error {
	instance, err := state.Load(opts.InstanceName)
	if err != nil {
		return err
	}

	resolvedRemote := opts.RemotePath
	if opts.RepoName != "" {
		found := false
		for _, r := range instance.Repos {
			if r.Name == opts.RepoName {
				found = true
				break
			}
		}
		if !found {
			return run.Errorf("'%s' isn't on '%s'. On it: %s", opts.RepoName, opts.InstanceName, reposList(instance))
		}
		if !strings.HasPrefix(opts.RemotePath, "/") {
			resolvedRemote = fmt.Sprintf("%s/%s/%s", config.RemoteCodeDir, opts.RepoName, opts.RemotePath)
		}
	}

	resolvedLocal := opts.LocalPath
	if resolvedLocal == "" {
		cwd, err := os.Getwd()
		if err != nil {
			return err
		}
		resolvedLocal = filepath.Join(cwd, path.Base(opts.RemotePath))
	}

	if err := sshutil.CopyFrom(instance.IP, resolvedRemote, resolvedLocal, instance.LoginKeyPath); err != nil {
		return err
	}
	fmt.Printf("pulled %s -> %s\n", resolvedRemote, resolvedLocal)
	return nil
}
